core/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/kubernetesModels.scala

```diff
@@ -0,0 +1,106 @@
+package org.broadinstitute.dsde.workbench.leonardo
+
+import enumeratum.{Enum, EnumEntry}
+import org.broadinstitute.dsde.workbench.google2.GKEModels.{KubernetesClusterId, KubernetesClusterName, NodepoolName}
+import org.broadinstitute.dsde.workbench.google2.KubernetesModels.KubernetesApiServerIp
+import org.broadinstitute.dsde.workbench.google2.KubernetesSerializableName.KubernetesNamespaceName
+import org.broadinstitute.dsde.workbench.google2.{Location, MachineTypeName, NetworkName, SubnetworkName}
+import org.broadinstitute.dsde.workbench.model.WorkbenchEmail
+import org.broadinstitute.dsde.workbench.model.google.GoogleProject
+
+/** Google Container Cluster statuses
+ *  see: https://cloud.google.com/kubernetes-engine/docs/reference/rest/v1/projects.locations.clusters#Cluster.Status
+ */
+
+case class KubernetesCluster(id: KubernetesClusterLeoId,
+                             googleProject: GoogleProject,
+                             clusterName: KubernetesClusterName,
+                             location: Location,
+                             status: KubernetesClusterStatus,
+                             serviceAccountInfo: WorkbenchEmail,
+                             samResourceId: KubernetesClusterSamResource,
+                             auditInfo: AuditInfo,
+                             asyncFields: KubernetesClusterAsyncFields,
+                             namespaces: Set[KubernetesNamespaceName],
+                             labels: LabelMap,
+                             nodepools: Set[Nodepool]
+                            )
+
+object KubernetesCluster {
+  implicit class EnrichedKubernetesCluster(cluster: KubernetesCluster) {
+    def getGKEClusterId: KubernetesClusterId = KubernetesClusterId(cluster.googleProject, cluster.location, cluster.clusterName)
+  }
+}
+
+case class SaveKubernetesCluster(googleProject: GoogleProject,
+                                 clusterName: KubernetesClusterName,
+                                 location: Location,
+                                 status: KubernetesClusterStatus,
+                                 serviceAccountInfo: WorkbenchEmail,
+                                 samResourceId: KubernetesClusterSamResource,
+                                 auditInfo: AuditInfo,
+                                 labels: LabelMap,
+                                 initialNodepool: Nodepool) //the clusterId specified here isn't used, and will be replaced by the id of cluster saved beforehand
+
+case class KubernetesClusterSamResource(resourceId: String)
+
+case class KubernetesClusterAsyncFields(apiServerIp: Option[KubernetesApiServerIp],
```

Comment:
Yeah, the flow will be

1. Front Leo inserts initial record, `networkInfo` and `apiServerIp` aren't specified
2. Back Leo creates the network info and submits the call to google. Now `networkInfo` is specified and `apiServerIp` is pending the creation in google
3. finally, once the creation is complete in google, both `networkFields` and `apiServerIp` will be specified

Reply:
can we make wait until 2,3 to finish before we update DB?...

Reply:
Hmm.. It's possible, but since they are discrete pollable Operations, I don't see why we wouldn't update them separately as they come available?

I'm open to doing them together and changing to `Option[AsyncFields]` to be consistent with our `createRuntime` and `CreateRuntimeResponse`, just curious why we chose to bunch all the db updates for the end of creation instead of doing them as each pollable action finishes.

Reply:
my thoughts on this is
1. I don't like we're making it 4 different permutations of just 2 fields. If we're going to add yet another field, it'll be 9 different presentations. This makes things difficult for caller of this data type because they need to accommodate all cases.
2. As far as I can see, there's not much point having say networkInfo persistted, but apiServerIp failed to create. We should probably clean things up if this happens, so why bother persisting networkInfo in the first case.